### PR TITLE
final eval

### PR DESCRIPTION
**Pull request by crewtool**
# Description

Neptune runs (observe stdout for randomly chose tokens):
- [1 GPU / llm-random](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=standard-view&detailsTab=charts&shortId=LLMRANDOM-20685&type=run&chartsFilter=final_eval)
-  [2 GPUs / llm-random](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=standard-view&detailsTab=charts&shortId=LLMRANDOM-20686&type=run&chartsFilter=final_eval)
- [4 GPUs / llm-random ](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=standard-view&detailsTab=charts&shortId=LLMRANDOM-20684&type=run&chartsFilter=final_eval)

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/argparse.py [hunk 1]
```diff
@@ -80,6 +80,20 @@ def introduce_parser_arguments(
     parser.add_argument("--train_dataset_path", type=str, default=None)
     parser.add_argument("--validation_dataset_path", type=str, default=None)
 
+    parser.add_argument("--final_eval_seed", type=int, default=1912)
+    parser.add_argument(
+        "--final_eval_dataloader_batch_size",
+        type=int,
+        default=64,
+        help="Batch size for final evaluation dataloader. It cannot be larger than the batch_size per GPU for the biggest model.",
```
Comment (j321m) [inline research/conditional/utils/argparse.py:88]: I'm a little confused by the biggest model. Why does the model size matter here? Do you mean the biggest batch_size per GPU in the experiment grid?

File: research/conditional/utils/conditional_trainer.py [hunk 2]
```diff
@@ -134,6 +137,25 @@ def _after_step_operations(self, step):
...
+    ):
+        if self.current_step == n_steps:
+            batches = [
+                self.final_eval_dataloader.get_batch()
+                for _ in range(
+                    self.n_sequences_final_eval // self.final_eval_dataloader_batch_size
+                )
+            ]
+
+            self._eval_single_variant(
```
Comment (j321m) [inline research/conditional/utils/conditional_trainer.py:152]: It seems like the function `_final_eval`() admits n_steps as argument, for the option of possibly making eval in the middle of the training (if `n_steps` arument `!= args.n_steps`).` If so, `varinat_name` an `stage` arguments schould be different. I suggest either adding them as arguments to `_final_eval` or deleting `n_steps` argument and taking it from `args`.

File: research/conditional/utils/argparse.py [hunk 3]
```diff
@@ -80,6 +80,20 @@ def introduce_parser_arguments(
...
+    )
+    parser.add_argument(
+        "--n_sequences_final_eval",
```
Comment (j321m) [inline research/conditional/utils/argparse.py:91]: I'm not sure if it's a good idea, but why don't we just specify `n_sequences_final_eval` and then choose `final_eval_dataloader_batch_size` dynamically as the biggest number that fits some constraints? We may be inefficient if we pick weird sequence number (like large prime naumber or something stupid), but it won't change the result.

**Comment by j321m** [comment]
Is it possible to show experimentally that eval batches in 2 different trainings are really the same?

**Comment by crewtool** [comment]
Short note for the record: We had the discussion as a team. Currently our dataloader works as follows:
We have following setup:
Two dataloaders A and B with the same seed, but A with batch size 48 and B with batch size 64.
We take first batch from both A and B. 
First 48 rows are exactly the same, but dataloader A have no way of getting rows 49-63 from dataloader B.
The situation is the same in subsequent batches. 
Therefore we have two choices:
- to have a dataloader with bsz such that the biggest model can process (smaller models obviously will do it as well)
- to have a dataloader with huge bsz and split it into smaller chunks so it fits model, but we need to adjust it for different models properly and also make sure each GPU uses proper chunk.

File: research/conditional/utils/argparse.py (hunk 3 above)
Comment (crewtool) [reply research/conditional/utils/argparse.py:91]: Discussed offline. See [this](https://github.com/llm-random/llm-random/pull/466#issuecomment-2457250987) comment.

File: research/conditional/utils/conditional_trainer.py (hunk 2 above)
Comment (crewtool) [reply research/conditional/utils/conditional_trainer.py:152]: We do not pass `args` to this funciton. `n_steps` is only for check if it is time for final eval

File: research/conditional/utils/conditional_trainer.py [hunk 4]
```diff
@@ -136,6 +140,31 @@ def _after_step_operations(self, step):
...
+    ):
+        if self.current_step == n_steps:
+            self.model.eval()
+            self.gradient_accumulation_steps = 1
+            losses = []
+            for _ in range(self.n_final_eval_batches):
+                batch = self.final_eval_dataloader.get_batch()
+                with torch.no_grad():
+                    loss, _ = self.calculate_loss_and_gradient(
+                        batch, self.batch_size // self.gradient_accumulation_steps
```
Comment (ludziej) [inline research/conditional/utils/conditional_trainer.py:155]: I don't really get why we divide here by self.gradient_accumulation_steps, while it's also hardcoded to be one. Why does it have to be one as well?

**Comment by ludziej** [comment]
Please add neptune experiments with current code, with final eval loss similar to train loss (because I'm still not entirely sure if it's correctly averaged).

**Comment by crewtool** [comment]
Neptune experiments updated

**Review by j321m — APPROVED** [review]

File: research/conditional/utils/check_args.py [hunk 5]
```diff
@@ -60,6 +60,10 @@ def check_args(args):
             new_scheduler_trapezoidal_slides.append(slide)
         args.scheduler_trapezoidal_slides = new_scheduler_trapezoidal_slides
 
+    assert (
```
[the remark this answers is not in the document]
Comment (crewtool) [reply research/conditional/utils/check_args.py:63]: Sure, nice idea!
